src/index.test.ts

```diff
@@ -0,0 +1,97 @@
+import { describe, test, beforeEach } from "node:test";
+import assert from "node:assert/strict";
+import { promisify } from "util";
+import { exec as lameExec } from "child_process";
+const exec = promisify(lameExec);
+import { Keypair } from "@solana/web3.js";
+import { getKeypairFromFile, getKeypairFromEnvironment } from "./index";
+import { writeFile } from "node:fs/promises";
+import base58 from "bs58";
+
+const log = console.log;
+
+const TEMP_DIR = "src/temp";
+
+describe("getKeypairFromFile", () => {
+  test("getting a keypair from a file", async () => {
+    const TEST_FILE_NAME = `${TEMP_DIR}/test-keyfile-do-not-use.json`;
+    const { stdout, stderr } = await exec(
+      `solana-keygen new --force --no-bip39-passphrase -o ${TEST_FILE_NAME}`,
+    );
+    assert(stdout.includes("Wrote new keypair"));
+    const keypair = await getKeypairFromFile(TEST_FILE_NAME);
+  });
+
+  test("throws a nice error if the file is missing", async () => {
+    try {
+      await getKeypairFromFile("I DONT EXIST");
+    } catch (thrownObject) {
+      const error = thrownObject as Error;
+      assert.strictEqual(
+        error.message,
+        `Could not read keypair from file at 'I DONT EXIST'`,
+      );
+    }
+  });
+
+  test("throws a nice error if the file is corrupt", async () => {
+    const TEST_FILE_NAME = `${TEMP_DIR}/corrupt-keyfile-do-not-use.json`;
+    await writeFile(TEST_FILE_NAME, "I AM CORRUPT");
+    try {
+      await getKeypairFromFile(TEST_FILE_NAME);
+    } catch (thrownObject) {
+      const error = thrownObject as Error;
+      assert.strictEqual(
+        error.message,
+        `Invalid secret key file at '${TEST_FILE_NAME}'!`,
+      );
+    }
+  });
+});
+
+describe("getKeypairFromEnvironment", () => {
+  test("getting a keypair from an environment variable (array of numbers format)", async () => {
+    const randomKeypair = Keypair.generate();
+    const TEST_ENV_VAR = "TEST_ENV_VAR";
+    process.env[TEST_ENV_VAR] = JSON.stringify(
+      Object.values(randomKeypair.secretKey),
+    );
```

Comment:
In general, set up preconditions in a setup hook.

```ts
describe("getKeypairFromEnvironment", () => {
  describe('given there is a keypair in an environment variable', () => {
    let expectedKeypair;
    beforeEach(() => {
      const expectedKeypair = Keypair.generate();
      process.env.TEST_ENV_VAR = JSON.stringify(
        Object.values(expectedKeypair.secretKey),
      );
    });
    test("getting a keypair from an environment variable (array of numbers format)", async () => {
      const actualKeypair = await getKeypairFromEnvironment(TEST_ENV_VAR);
      assert.equal(actualKeypair, expectedKeypair);
    });
  });
});
```

This way your test body only tests the _actual_ code under test.

Reply:
You can apply this technique everywhere else there's a precondition to materialize.

Reply:
Ack - I did actually have a beforeEach(), however the preconditions are unique to each test so I got rid of it. 

I do see what you're saying though - a failure should only be for what's being tested, say if writing a file fails, we want that to show as an error in the beforeAll(). I'll tweak this - thanks!

Reply:
Done!